File: rpi/acurite523.py
#!/usr/bin/python3 -u

"""
Parses 433MHz RF signals from the AcuRite 00523M refrigerator/freezer units.
"""

import struct

SIGNAL_BIT_LENGTH = 48              # Length of bitstream in bits
SIGNAL_INV = -2
SIGNAL_BIT_0_OFF = 0
SIGNAL_BIT_0_ON = 1
SIGNAL_BIT_1_OFF = 2
SIGNAL_BIT_1_ON = 3
SIGNAL_BITSTREAM_OFF = 4
SIGNAL_BITSTREAM_ON = 5
SIGNAL_CHUNK_END = 6
ACURITE523_SIG_FREEZER = 0xc049  # Signatures seem to be hardcoded?
ACURITE523_SIG_FRIDGE = 0xc07c
MODEL_ACURITE523 = 1592
MODEL_ACURITE609 = 6585
DEVICE_FREEZER   = 9690
DEVICE_FRIDGE    = 7784
TAG_TEMPMONITOR = 0x38073162

class Acurite523():
    class Model():
        """Parsing and chunk-building for model-specific RF signals.
        """
        def __init__(self, devices):
            self.devices = devices
            self.bitstream = 0       # All bits received in a single bitstream
            self.bitstream_size = 0  # Size in bits of current bitstream
            self.last_rfs_type = SIGNAL_INV
            self.bitstream_open = False
            self.chunk_open = False
            self.bitstream_opener_count = 0

        def clear(self):
            self.bitstream = 0
            self.bitstream_size = 0
            self.last_rfs_type = SIGNAL_INV
            self.bitstream_open = False
            self.chunk_open = False
            self.bitstream_opener_count = 0

        def get_rfs_type(self, rfs, duration):
            """Returns the type of RF signal received.
            :param int rfs: RF signal received; either 0 or 1
            :param int duration: signal duration, in microseconds
            :return: the signal type
            :rtype: int
            """
            if rfs == 0:
                if duration >= 100 and duration < 300:
                    return SIGNAL_BIT_0_OFF
                elif duration >= 300 and duration < 500:
                    return SIGNAL_BIT_1_OFF
                elif duration >= 500 and duration < 700:
                    return SIGNAL_BITSTREAM_OFF
            elif rfs == 1:
                if duration >= 100 and duration < 300:
                    return SIGNAL_BIT_1_ON
                elif duration >= 300 and duration < 500:
                    return SIGNAL_BIT_0_ON
                elif duration >= 500 and duration < 700:
                    return SIGNAL_BITSTREAM_ON
                elif duration >= 20000 and duration < 60000:
                    return SIGNAL_CHUNK_END
            return SIGNAL_INV

        def is_bit_signal(self, rfs_type):
            return (
                    rfs_type == SIGNAL_BIT_0_OFF or 
                    rfs_type == SIGNAL_BIT_0_ON or 
                    rfs_type == SIGNAL_BIT_1_OFF or 
                    rfs_type == SIGNAL_BIT_1_ON)

        def open_bitstream(self):
            self.bitstream_open = True
            self.bitstream_size = 0
            self.bitstream = 0

        def close_bitstream(self):
            self.bitstream_open = False
            self.bitstream_size = 0
            self.bitstream = 0

        def open_chunk(self):
            self.chunk_open = True
            self.open_bitstream()

        def close_chunk(self):
            self.chunk_open = False
            self.close_bitstream()

        def parse_rf(self, duration, rfs):
            """Parse a single RF signal and update chunk/bitstreams.
            """
            result = 0
            rfs_type = SIGNAL_INV
            rfs_type = self.get_rfs_type(rfs, duration)
            if self.last_rfs_type == SIGNAL_BITSTREAM_OFF or not self.chunk_open:
                if rfs_type == SIGNAL_BITSTREAM_ON:
                    self.bitstream_opener_count += 1
                if self.bitstream_opener_count == 4:
                    self.bitstream_opener_count = 0
                    if not self.chunk_open:
                        # Chunk start
                        self.open_chunk()
                    # Bitstream start
            elif self.last_rfs_type == SIGNAL_BIT_0_OFF and self.chunk_open:
                if rfs_type == SIGNAL_BIT_0_ON and self.bitstream_size < SIGNAL_BIT_LENGTH:
                    self.bitstream_size += 1
                    if self.bitstream_size == SIGNAL_BIT_LENGTH:
                        result = self.bitstream
                        self.close_bitstream()
                elif rfs_type == SIGNAL_BIT_1_ON and self.bitstream_size == SIGNAL_BIT_LENGTH:
                    # Bitstream end
                    result = self.bitstream
                    self.close_bitstream()
                elif rfs_type == SIGNAL_CHUNK_END:
                    # Chunk end
                    if self.bitstream_size == SIGNAL_BIT_LENGTH:
                        result = self.bitstream
                    self.close_chunk()
                self.bitstream_opener_count = 0
            elif self.last_rfs_type == SIGNAL_BIT_1_OFF and self.chunk_open:
                if rfs_type == SIGNAL_BIT_1_ON and self.bitstream_size < SIGNAL_BIT_LENGTH:
                    self.bitstream |= (1 << (SIGNAL_BIT_LENGTH - self.bitstream_size - 1))
                    self.bitstream_size += 1
                    if self.bitstream_size == SIGNAL_BIT_LENGTH:
                        result = self.bitstream
                        self.close_bitstream()
            self.last_rfs_type = rfs_type
            # Done
            return result

    class Device():
        def __init__(self, device, verbosity=0):
            self.print_verbose = print if verbosity > 1 else lambda *a, **k: None
            self.print_debug = print if verbosity > 2 else lambda *a, **k: None
            self.device = device
            self.temperature = None
            self.battery = -1
            if device == DEVICE_FREEZER:
                self.signature = ACURITE523_SIG_FREEZER
            elif device == DEVICE_FRIDGE:
                self.signature = ACURITE523_SIG_FRIDGE
            else:
                self.signature = 0

        def create_payload(self, status):
            return struct.pack('<IHHbbhh', TAG_TEMPMONITOR, 
                    MODEL_ACURITE523, self.device, status, self.battery, 
                    int(self.temperature * 10), 0)

        def validate_checksum(self, bitstream):
            checksum = bitstream & 0xff
            calculated = (((bitstream >> 8) & 0xff) + 
                    ((bitstream >> 16) & 0xff) +
                    ((bitstream >> 24) & 0xff) + 
                    ((bitstream >> 32) & 0xff) + 
                    ((bitstream >> 40))) & 0xff
            if checksum != calculated:
                self.print_verbose(f'bad checksum: {calculated:#x}', end='')
            return checksum == calculated

        def validate_bitstream(self, bitstream):
            """Validates the signature and checksum of the specified bitstream.

            :param int bitstream: the bitstream of data to validate
            :return: True if bitstream is good, False if bad
            """
            if bitstream == 0:
                return False
            sig = bitstream >> 32
            if sig != self.signature:
                return False
            self.print_verbose(f'[{bitstream:048b}] ', end='')
            if not self.validate_checksum(bitstream):
                self.print_verbose(f', signature: {sig:#x}')
                return False
            bat = (bitstream >> 30) & 0x03
            temp = ((bitstream >> 9) & 0x3f80) | ((bitstream >> 8) & 0x7f)
            temp = (temp - 1800) / 18
            if temp < -40 and temp >= 70:
                self.print_verbose(f'invalid data {temp:.1f}C')
                return False
            # Set the instance values
            self.battery = bat
            self.temperature = temp
            temp = temp * 1.8 + 32
            if self.signature == ACURITE523_SIG_FREEZER:
                self.print_verbose('freezer: ', end='')
            elif self.signature == ACURITE523_SIG_FRIDGE:
                self.print_verbose('fridge: ', end='')
            self.print_verbose(f'{temp:.1f}F, battery={bat}')
            return True

File: esp32/acurite523.cpp
#include "tempstation.h"

/**
 * Parsing && chunk-building for model-specific RF signals.
 */
Acurite523::Model::Model(std::vector<Acurite523::Device> devices) {
    this->devices = devices;
}

void Acurite523::Model::clear() {
    this->bitstream = 0;      // All bits received in a single bitstream
    this->bitstream_size = 0; // Size in bits of current bitstream
    this->last_rfs_type = ACURITE523_SIGNAL_INV;
    this->bitstream_open = false;
    this->bitstream_opener_count = 0;
    // Do not reset chunk variables
}

int Acurite523::Model::get_rfs_type(uint8_t rfs, uint32_t duration) {
    /* Returns the type of RF signal received.
       :param int rfs: RF signal received; either 0 || 1
       :param int duration: signal duration, in microseconds
       :return: the signal type
       :rtype: int
       */
    if (rfs == 0) {
        if (duration >= 100 && duration < 300)
            return ACURITE523_SIGNAL_BIT_0_OFF;
        else if (duration >= 300 && duration < 500)
            return ACURITE523_SIGNAL_BIT_1_OFF;
        else if (duration >= 500 && duration < 700)
            return ACURITE523_SIGNAL_BITSTREAM_OFF;
    }
    else if (rfs == 1) {
        if (duration >= 100 && duration < 300)
            return ACURITE523_SIGNAL_BIT_1_ON;
        else if (duration >= 300 && duration < 500)
            return ACURITE523_SIGNAL_BIT_0_ON;
        else if (duration >= 500 && duration < 700)
            return ACURITE523_SIGNAL_BITSTREAM_ON;
        else if (duration >= 20000 && duration < 60000)
            return ACURITE523_SIGNAL_CHUNK_END;
    }
    return ACURITE523_SIGNAL_INV;
}

bool Acurite523::Model::is_bit_signal(int rfs_type) {
    return 
        rfs_type == ACURITE523_SIGNAL_BIT_0_OFF || 
        rfs_type == ACURITE523_SIGNAL_BIT_0_ON || 
        rfs_type == ACURITE523_SIGNAL_BIT_1_OFF || 
        rfs_type == ACURITE523_SIGNAL_BIT_1_ON;
}

void Acurite523::Model::open_bitstream() {
    bitstream_open = true;
    bitstream_size = 0;
    bitstream = 0;
}

void Acurite523::Model::close_bitstream() {
    bitstream_open = false;
    bitstream_size = 0;
    bitstream = 0;
}

void Acurite523::Model::open_chunk() {
    chunk_open = true;
    open_bitstream();
}

void Acurite523::Model::close_chunk() {
    chunk_open = false;
    close_bitstream();
}

uint64_t Acurite523::Model::parse_rf(uint32_t duration, uint8_t rfs) {
    /* Parse a single RF signal && update chunk/bitstreams.
    */
    uint64_t result = 0;
    int rfs_type = ACURITE523_SIGNAL_INV;
    rfs_type = get_rfs_type(rfs, duration);
    if (last_rfs_type == ACURITE523_SIGNAL_BITSTREAM_OFF || !chunk_open) {
        if (rfs_type == ACURITE523_SIGNAL_BITSTREAM_ON)
            bitstream_opener_count += 1;
        if (bitstream_opener_count == 4) {
            bitstream_opener_count = 0;
            if (!chunk_open)
                open_chunk();
        }
    }
    else if (last_rfs_type == ACURITE523_SIGNAL_BIT_0_OFF && chunk_open) {
        if (rfs_type == ACURITE523_SIGNAL_BIT_0_ON && bitstream_size < ACURITE523_SIGNAL_BIT_LENGTH) {
            bitstream_size += 1;
            if (bitstream_size == ACURITE523_SIGNAL_BIT_LENGTH) {
                result = bitstream;
                close_bitstream();
            }
        }
        else if (rfs_type == ACURITE523_SIGNAL_BIT_1_ON && bitstream_size == ACURITE523_SIGNAL_BIT_LENGTH) {
            // Bitstream end
            result = bitstream;
            close_bitstream();
        }
        else if (rfs_type == ACURITE523_SIGNAL_CHUNK_END) {
            // Chunk end
            if (bitstream_size == ACURITE523_SIGNAL_BIT_LENGTH)
                result = bitstream;
            close_chunk();
        }
        bitstream_opener_count = 0;
    }
    else if (last_rfs_type == ACURITE523_SIGNAL_BIT_1_OFF && chunk_open) {
        if (rfs_type == ACURITE523_SIGNAL_BIT_1_ON && bitstream_size < ACURITE523_SIGNAL_BIT_LENGTH) {
            bitstream |= ((uint64_t)1L << (ACURITE523_SIGNAL_BIT_LENGTH - bitstream_size - 1));
            bitstream_size += 1;
            if (bitstream_size == ACURITE523_SIGNAL_BIT_LENGTH) {
                result = bitstream;
                close_bitstream();
            }
        }
    }
    last_rfs_type = rfs_type;

    // Done
    return result;
}

Acurite523::Device::Device(uint16_t device) {
    this->device = device;
    this->temperature = 0;
    this->battery = 0;
    // Signatures are hardcoded for these devices
    if (device == DEVICE_FREEZER)
        this->signature = ACURITE523_SIG_FREEZER;
    else if (device == DEVICE_FRIDGE)
        this->signature = ACURITE523_SIG_FRIDGE;
    else
        this->signature = 0;
}

Payload *Acurite523::Device::create_payload(uint8_t status) {
    Payload *payload = new Payload();
    payload->tag = TAG_TEMPMONITOR;
    payload->model = MODEL_ACURITE523;
    payload->device = device;
    payload->status = status;
    payload->battery = battery;
    payload->temperature = int16_t(temperature * 10);
    payload->humidity = 0;
    return payload;
}

bool Acurite523::Device::validate_checksum(uint64_t bitstream) {
    uint8_t checksum = bitstream & 0xff;
    uint32_t calculated = (((bitstream >> 8) & 0xff) + 
            ((bitstream >> 16) & 0xff) +
            ((bitstream >> 24) & 0xff) + 
            ((bitstream >> 32) & 0xff) + 
            ((bitstream >> 40))) & 0xff;
    if (checksum != calculated) {
        Serial.print("bad checksum: ");
        Serial.print(checksum, HEX);
    }
    return checksum == calculated;
}

bool Acurite523::Device::validate_parity(uint8_t parity, uint8_t value) {
    int on_bits = 0;
    for (int i = 0; i < 8; i++) {
        on_bits += value & 1;
        value >>= 1;
    }
    return (on_bits % 2) == parity;
}

bool Acurite523::Device::validate_bitstream(uint64_t bitstream) {
    /* Validates the signature and checksum of the specified bitstream.

       :param int bitstream: the bitstream of data to validate
       :return: true if bitstream is good, false if bad
       */
    // Parse and validate data
    if (bitstream == 0)
        return false;
    uint16_t sig = bitstream >> 32;
    if (sig != signature)
        return false;
    Serial.print("[");
    Serial.print(bitstream, BIN);
    Serial.print("] ");
    if (!validate_checksum(bitstream)) {
        Serial.print(", signature: ");
        Serial.println(signature, HEX);
        return false;
    }
    uint8_t bat = (bitstream >> 30) & 0x03;
    // Validate parity bit
    uint8_t parity1 = (bitstream >> 15) & 1;
    uint8_t byte1 = (bitstream >> 8) & 0x7f;
    uint8_t parity2 = (bitstream >> 23) & 1;
    uint8_t byte2 = (bitstream >> 16) & 0x7f;
    if (!validate_parity(parity1, byte1) || !validate_parity(parity2, byte2)) {
        Serial.println("parity bit fail");
        return false;
    }
    // Validate temperature
    float temp = ((uint16_t)byte2 << 7) | byte1;
    temp = (temp - 1800) / 18;
    if (temp < -40 || temp >= 70) {
        Serial.print("invalid data: ");
        Serial.print(temp);
        Serial.println("C");
        return false;
    }
    // Set the instance values
    battery = bat;
    temperature = temp;
    if (signature == ACURITE523_SIG_FREEZER)
        Serial.print("freezer: ");
    else if (signature == ACURITE523_SIG_FRIDGE)
        Serial.print("fridge: ");
    Serial.print(temp * 1.8 + 32);
    Serial.print("F, battery=");
    Serial.println(bat);
    return true;
}

